server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java

```diff
@@ -294,17 +314,38 @@ private boolean onWorkerResourceStatus(final WorkerResourceStatus workerEvent)
 
 
     /**
+     * Handles a {@link WorkerHeartbeat} event.
+     *
      * Assumptions:
-     * 1. Heartbeats from workers of terminated jobs are ignored at a higher level
-     * @param workerEvent
-     * @throws InvalidWorkerStateChangeException
+     *
+     * 1. Heartbeats from workers of terminated jobs are ignored at a higher level.
+     *
+     * This method will not move a worker to the {@link WorkerState#Started} state if the worker fails validation.
+     * Because of this, the worker's state won't get persisted. This means the worker should be stuck in
+     * the {@link WorkerState#Accepted} state which should eventually be caught and resubmitted.
+     *
+     * @param workerEvent a {@link WorkerHeartbeat} event.
+     *
+     * @throws InvalidWorkerStateChangeException if it fails to persist worker state
+     *         via {@link JobWorker#setState(WorkerState, long, JobCompletedReason)}.
+     * @throws MissingWorkerPortsException if it has no {@link WorkerPorts} from the metadata or if the worker ports
+     *         don't pass its validation.
      */
-    private boolean onHeartBeat(WorkerHeartbeat workerEvent) throws InvalidWorkerStateChangeException {
+    private boolean onHeartBeat(WorkerHeartbeat workerEvent)
+            throws InvalidWorkerStateChangeException, MissingWorkerPortsException {
+
         numHeartBeatsReceived.increment();
         if (LOGGER.isTraceEnabled()) {
             LOGGER.trace("Job {} Processing onHeartBeat for {}", this.metadata.getJobId(),
                 metadata.getWorkerId());
         }
+
+        WorkerPorts workerPorts = metadata.getPorts().orElseThrow(MissingWorkerPortsException::new);
```

Comment:
Need a closer review here. Idea is to make sure worker metadata is good even on a heartbeat. If not, bubble up this exception.

Reply:
feels odd to have this check here. The validation should ideally happen on state changes at a central place. A worker goes from `Accepted` to `Launched` (only when there are valid ports after scheduling) to `StartInitiated` (worker lifecycle start) to `Started` (when there is a heartbeat). Are we seeing a HB before the worker is in `Launched` state?
If we can ensure the WorkerMetadata is in a consistent state with valid port info if it is in any of Launched/StartInitiated/Started states at the time of transitioning to those states, we might be able to avoid these validity checks at multiple places.

Reply:
Agreed. I wasn't sure about the HB sending. Let me remove this and test out the HB and state transitions.

Reply:
To clarify, originally Neeraj and I were thinking to catch it here but it's worth testing out situation you pointed out.

Reply:
Thinking this through some more, I think we can do this without going through heartbeats. in `JobActor#initializeRunningWorkers` https://github.com/Netflix/mantis-control-plane/pull/26/files#diff-81433f6ab0cca184eca4716606d9b3eeL1363, we can just check for missing ports without interfering with the fake heartbeat send. Then, we can get rid of the exception handling. cc @neerajrj